lib/crawler.js

```diff
@@ -98,46 +100,89 @@ Crawler.prototype.init = function init (options) {
 	// Don't make these options persist to individual queries
 	self.globalOnlyOptions = ['maxConnections', 'rateLimit', 'priorityRange', 'homogeneous', 'skipDuplicates', 'rotateUA'];
 
-	self.limiters = new Bottleneck.Cluster(self.options.maxConnections,self.options.rateLimit,self.options.priorityRange, self.options.priority, self.options.homogeneous);
+	self.limiters = new Bottleneck.Cluster(self.options.maxConnections, self.options.rateLimit, self.options.priorityRange, self.options.priority, self.options.homogeneous);
+
+	//maintain the http2 sessions
+	self.http2Connections = {};
 
 	level = self.options.debug ? 'debug' : 'info';
 
-	if(self.options.logger)
+	if (self.options.logger)
 		log = self.options.logger.log.bind(self.options.logger);
 
 	self.log = log;
 
 	self.seen = new seenreq(self.options.seenreq);
-	self.seen.initialize().then(()=> log('debug', 'seenreq is initialized.')).catch(e => log('error', e));
+	self.seen.initialize().then(() => log('debug', 'seenreq is initialized.')).catch(e => log('error', e));
+
+	self.on('_release', function () {
+		log('debug', 'Queue size: %d', this.queueSize);
 
-	self.on('_release', function(){
-		log('debug','Queue size: %d',this.queueSize);
+		if (this.limiters.empty) {
 
-		if(this.limiters.empty)
+			if (Object.keys(self.http2Connections).length > 0) self._clearHttp2Session();
 			return this.emit('drain');
+		}
 	});
 };
 
-Crawler.prototype.setLimiterProperty = function setLimiterProperty (limiter, property, value) {
+Crawler.prototype.setLimiterProperty = function setLimiterProperty(limiter, property, value) {
 	var self = this;
 
-	switch(property) {
-	case 'rateLimit': self.limiters.key(limiter).setRateLimit(value);break;
-	default: break;
+	switch (property) {
+		case 'rateLimit': self.limiters.key(limiter).setRateLimit(value); break;
+		default: break;
 	}
 };
 
-Crawler.prototype._inject = function _inject (response, options, callback) {
+
+Crawler.prototype.generateHttp2RequestLine = function (options) {
+	const requestLine = {};
+	const urlObj = new URL(options.uri);
+	requestLine[':method'] = options.method || 'POST';
+	requestLine[':path'] = urlObj.pathname;
+	requestLine[':scheme'] = urlObj.protocol.replace(':', '');
+	requestLine[':authority'] = urlObj.hostname;
+
+	return requestLine;
+};
+
+Crawler.prototype.generateHttp2RequestBody = function (options) {
+	if (options.headers[':method'] === 'GET') return null;
+
+	let data = '';
```

Comment:
what should be default value if there is no body in a POST request according to RFC?